dbms/src/Storages/DeltaMerge/VersionChain/DMFileHandleIndex.h

```diff
@@ -0,0 +1,241 @@
+    }
+
+    // Find the base(oldest) version of the given handle in the given pack.
+    std::optional<RowID> getBaseVersion(const DMContext & dm_context, HandleRefType h, UInt32 clipped_pack_id)
+    {
+        loadHandleIfNotLoaded(dm_context);
+        ColumnView<HandleType> handle_col(*clipped_handle_packs[clipped_pack_id]);
+        auto itr = std::lower_bound(handle_col.begin(), handle_col.end(), h);
+        if (itr != handle_col.end() && *itr == h)
+            return itr - handle_col.begin() + clipped_pack_offsets[clipped_pack_id];
+        return {};
+    }
+
+    std::vector<HandleType> loadPackIndex(const DMContext & dm_context) const
+    {
+        auto max_values = loadPackMaxValue<HandleType>(dm_context, *dmfile, MutSup::extra_handle_id);
+        return std::vector<HandleType>(
+            max_values.begin() + clipped_pack_range.start_pack_id,
+            max_values.begin() + clipped_pack_range.end_pack_id);
+    }
+
+    std::vector<UInt32> loadPackOffsets() const
+    {
+        const auto & pack_stats = dmfile->getPackStats();
+        std::vector<UInt32> pack_offsets(clipped_pack_range.count(), 0);
+        for (UInt32 clipped_pack_id = 1; clipped_pack_id < pack_offsets.size(); ++clipped_pack_id)
+        {
+            UInt32 real_pack_id = clipped_pack_id + clipped_pack_range.start_pack_id;
+            pack_offsets[clipped_pack_id] = pack_offsets[clipped_pack_id - 1] + pack_stats[real_pack_id - 1].rows;
+        }
+        return pack_offsets;
+    }
+
+    void loadHandleIfNotLoaded(const DMContext & dm_context)
+    {
+        if (likely(clipped_need_read_packs.empty()))
+            return;
+
+        auto read_pack_ids = std::make_shared<IdSet>();
+        for (UInt32 i = 0; i < clipped_need_read_packs.size(); ++i)
+        {
+            if (clipped_need_read_packs[i])
+                read_pack_ids->insert(i + clipped_pack_range.start_pack_id);
+        }
+
+        auto pack_filter = DMFilePackFilter::loadFrom(
+            dm_context,
+            dmfile,
+            /*set_cache_if_miss*/ true,
+            /*rowkey_ranges*/ {}, // Empty means all
+            EMPTY_RS_OPERATOR,
+            read_pack_ids);
+
+        DMFileReader reader(
+            dmfile,
+            {getHandleColumnDefine<HandleType>()},
+            isCommonHandle<HandleType>(),
+            /*enable_handle_clean_read*/ false,
+            /*enable_del_clean_read*/ false,
+            /*is_fast_scan*/ false,
+            /*max_data_version*/ std::numeric_limits<UInt64>::max(),
+            std::move(pack_filter),
+            dm_context.global_context.getMarkCache(),
+            /*enable_column_cache*/ false,
+            /*column_cache*/ nullptr,
+            dm_context.global_context.getSettingsRef().max_read_buffer_size,
+            dm_context.global_context.getFileProvider(),
+            dm_context.global_context.getReadLimiter(),
+            DEFAULT_MERGE_BLOCK_SIZE,
+            /*read_one_pack_every_time*/ true,
```

Comment:
Maybe we can use `DMFileReader::ReadBlockInfo` to fill `clipped_handle_packs`, rather than set `read_one_pack_every_time`.